File: dgpt/whatif.py
"""The drop-in kit: what the 2027 what-if tab needs to add a player to the field.

The site already has one what-if — `docs/js/sim.js`, which replays the 2026
cutline for a player who exists. This is the other question, and it is only
askable about a season nobody has played: *drop a rating into the 2027 field,
send it to some number of events, and where does it come out?*

Answering that in the browser needs two things the projection bundle does not
otherwise carry:

  the field's own ladder   the season totals a hypothetical player is ranked
                           against, per simulated season and per rank
  the field's shape        who is standing on each tee, and what a place is
                           worth there

Both are summaries, not raw simulation output. The whole kit is about 25 KB
against a 155 KB bundle, which is what makes the tab a fetch rather than a
service.

## The ladder, and why quantiles are enough

The honest object is a (n_sims x n_ranks) matrix: in each simulated season,
the total held by each standings position. At 20,000 seasons that is far too
big to ship, so what goes out is its *marginal* quantiles — for each rank, the
distribution of the total sitting at it.

The client reconstructs a season by drawing one uniform and reading every rank
off at that quantile. That is a comonotone approximation: it ties rank 1's
season to rank 50's more tightly than the simulation does. It is also exactly
right for every number the tab prints, because each one is a statement about a
single rank —

    P(rank <= r)  ==  P(my total > the r-th best total)

— and the r-th best total keeps its true marginal distribution under that
construction. What the approximation loses is joint structure across ranks,
which nothing here reads. Quantiles also preserve the ladder's ordering: if
L_r >= L_(r+1) in every season, the same holds at every quantile, so the
reconstructed ladder never crosses itself.

Ranks are shipped one at a time down to the histogram's depth and then
log-spaced to the bottom of the roster, so "you finish about 180th" stays a
sentence the tab can say without shipping 689 rungs.

## The performance paths

Two Cup doors are not thresholds at all: the top finishers at the first
playoff event advance to the second, and the top finishers at the second take
the Cup places the standings cut did not. Whether a given finish walks through
depends on how many already-qualified players finished ahead of it — a fact
about the whole field that a client-side draw cannot reconstruct.

So it is measured instead, where it actually happened: P(advances | finished
here, needed it), counted over every simulated season. The client draws a
place and reads the odds off.

## What stays approximate

The field is frozen. A player dropped into it takes places off everyone else,
which would move the ladder they are measured against, and the kit is built
from seasons they were not in. That is the same trade `sim.js` makes and the
same reason: correcting it needs a re-simulation of the field, which is not a
thing a slider can do between frames. The tab says so above the numbers.
"""
from __future__ import annotations

import numpy as np

# Quantile grid for every shipped distribution: 0, 1/50, ... 1. The client
# draws a uniform and interpolates, so the endpoints are the observed min and
# max and everything between is a 2%-wide band — finer than the tab's own
# display precision, which rounds above a tenth of a point.
LADDER_Q = 51

# Rating bands per event field. Twenty equal-weight bins describe a bimodal
# field well enough to rank against and cost 20 numbers; a single mean and
# standard deviation cost two and are exactly what `sim.js` had to stop using
# when an open field broke them (see its `drawPlace`).
BAND_N = 20

# Places covered by each class's exported points curve. Has to outrun the
# deepest 2027 field (the first playoff event fills to 120) for the same
# reason `export.CURVE_DEPTH` does: the client reads a drawn place straight
# off it and scores anything past the end as zero.
CURVE_DEPTH = 200


def ladder_ranks(depth: int, n: int, growth: float = 1.15) -> list[int]:
    """Standings positions the ladder is shipped at.

    Every rank through `depth` — that is the finishing histogram's own depth,
    so the tab can draw the same sparkline every other row on the page draws —
    and then geometrically to the bottom of the roster, which is what keeps
    "you finish about 180th" answerable without shipping 689 rungs.
    """
    top = min(depth, n)
    ranks = list(range(1, top + 1))
    r = top
    while r < n:
        r = min(n, max(r + 1, int(round(r * growth))))
        ranks.append(r)
    return ranks


def field_bands(ratings: np.ndarray, attendance: np.ndarray,
                nbins: int = BAND_N) -> list[float]:
    """The projected field of one event, as `nbins` equal-weight ratings.

    Attendance-weighted quantiles of the roster's ratings at the midpoint of
    each bin, so the list reads as "one nbins-th of the field is about this
    good" — which is how the client uses it, as a discrete opponent pool.
    """
    w = np.clip(np.asarray(attendance, dtype=float), 0.0, None)
    total = w.sum()
    if total <= 0:
        return []
    order = np.argsort(ratings)
    r, cw = np.asarray(ratings)[order], np.cumsum(w[order])
    targets = (np.arange(nbins) + 0.5) / nbins * total
    ix = np.clip(np.searchsorted(cw, targets), 0, r.size - 1)
    return [round(float(x), 1) for x in r[ix]]


def _quantiles(samples: np.ndarray) -> list[float]:
    """The shipped quantile grid of one series, rounded to a tenth of a point."""
    qs = np.quantile(samples, np.linspace(0.0, 1.0, LADDER_Q))
    return [round(float(x), 1) for x in qs]


class Kit:
    """Accumulates the drop-in summary alongside a projection run.

    One instance per projected season; `project._one_chunk` feeds it at the
    four points where the standings are ranked and the two where a performance
    path is decided. It holds one float32 row per simulated season per stream
    — about 6 MB at 20,000 seasons — and collapses to quantiles at the end.
    """

    def __init__(self, ranks: list[int], field_size: int,
                 perf_depth: tuple[int, int]) -> None:
        self.ranks = np.array(ranks, dtype=np.int64)
        # `ladder_ranks` ends at the roster, so the last rung is its size. Every
        # depth below is clamped to it, because the qualification constants are
        # written for a full tour and a division can be smaller than its own
        # Cup field — FPO's performance path already admits more players than
        # `tests/test_small_division.py` puts in the table.
        n = int(self.ranks[-1])
        self.field_size = min(field_size, n)
        self.play1_depth, self.play2_depth = (min(d, n) for d in perf_depth)
        self._ladder: list[np.ndarray] = []
        self._gates: dict[str, list[np.ndarray]] = {"play1": [], "play2": []}
        self._num = {"play2": np.zeros(self.play1_depth + 1),
                     "champ": np.zeros(self.play2_depth + 1)}
        self._den = {"play2": np.zeros(self.play1_depth + 1),
                     "champ": np.zeros(self.play2_depth + 1)}
        self._seed_rating = np.zeros(self.field_size)
        self._rows = 0

    # -- the three ranked moments in a simulated season ------------------

    def gate(self, which: str, totals: np.ndarray, rank: int) -> None:
        """The total sitting at `rank` — the line a playoff field is cut on."""
        k = min(rank, totals.shape[1]) - 1
        v = -np.partition(-totals, k, axis=1)[:, k]
        self._gates[which].append(v.astype(np.float32))

    def final(self, totals: np.ndarray, ratings: np.ndarray) -> None:
        """The finished standings: the ladder, and who is holding each seed."""
        order = np.argsort(-totals, axis=1)
        srt = np.take_along_axis(totals, order, axis=1)
        self._ladder.append(srt[:, self.ranks - 1].astype(np.float32))
        self._seed_rating += ratings[order[:, : self.field_size]].sum(axis=0)
        self._rows += totals.shape[0]

    # -- the two performance paths ---------------------------------------

    def perf(self, which: str, place: np.ndarray, eligible: np.ndarray,
             gained: np.ndarray) -> None:
        """Count a performance path by the finishing place that walked it."""
        depth = self._num[which].size - 1
        pl = np.clip(place, 1, depth)
        self._num[which] += np.bincount(pl[gained].ravel(), minlength=depth + 1)
        self._den[which] += np.bincount(pl[eligible].ravel(), minlength=depth + 1)

    def _perf_rates(self, which: str) -> list[float]:
        """P(walks the path | finished here), index 0 = first place.

        An empty bin is a place nobody who needed the path ever finished in,
        not evidence that finishing there stopped counting — so it carries a
        neighbour rather than a zero. The rate falls monotonically with place,
        so a gap in the middle takes the place above it and a gap at the top
        takes the best rate that was measured.

        The top of this curve is always a gap, and for a reason worth naming:
        winning the first playoff event banks enough to clear the second one's
        points cut outright, so a winner is never someone the performance path
        has to carry. Reading that as "a win is worth nothing" would be the
        one place a zero-fill could actually mislead the tab.
        """
        num, den = self._num[which], self._den[which]
        rates = []
        last = None
        for p in range(1, num.size):
            if den[p] > 0:
                last = num[p] / den[p]
            rates.append(last)
        head = next((r for r in rates if r is not None), 0.0)
        return [round(float(head if r is None else r), 4) for r in rates]

    # -- what goes in the bundle -----------------------------------------

    def build(self, curves: dict[str, list[float]], stroke_by_rank: list[int],
              invite: tuple[str, ...]) -> dict:
        """The `meta.whatif` block, whole.

        `curves`, `stroke_by_rank` and `invite` come from the caller because
        they are the projection's own points, seed and invitation tables —
        this module summarises a run, it does not get a second opinion about
        what a place is worth or which wins open a door.
        """
        ladder = np.concatenate(self._ladder, axis=0)
        return {
            "q": LADDER_Q,
            "ranks": [int(r) for r in self.ranks],
            # per rank, the quantiles of the season total sitting at it
            "ladder": [_quantiles(ladder[:, j]) for j in range(self.ranks.size)],
            # the lines the two playoff fields are cut on, same grid
            "play1_line": _quantiles(np.concatenate(self._gates["play1"])),
            "play2_line": _quantiles(np.concatenate(self._gates["play2"])),
            # P(advances) by finishing place, index 0 = first
            "play2_perf": self._perf_rates("play2"),
            "champ_perf": self._perf_rates("champ"),
            # mean rating of whoever finishes at each standings position, for
            # the Cup field a dropped-in player would tee off against
            "cup_field": [round(float(x), 1) for x in self._seed_rating / max(self._rows, 1)],
            "cup_strokes": list(stroke_by_rank),
            # classes whose winner takes the Cup's special invite
            "invite": list(invite),
            "curves": curves,
            "bands": BAND_N,
        }

File: docs/js/whatif.js
/* The What-if tab: a player who does not exist, dropped into the 2027 field.

   Every other view on this site reports on people who are really there. This
   one puts a rating and a schedule into the projected 2027 season and reads
   back the same columns the 2027 tab prints for everyone else — so the answer
   lands in a shape a reader has already learned.

   The arithmetic is `dropIn()` in `sim.js`, against the summary
   `dgpt/whatif.py` ships in `meta.whatif`. This module is the controls, the
   readout, and the caveats — which are not decoration here: the numbers below
   are a projection of a projection, and the page has to say so before it says
   anything else. */

import { $, agoText, fmtPct, fmtPctExact, fmtPts, fmtStroke, ordinal, state, updatedAt } from "./core.js";
import { tipAttrs } from "./tooltip.js";
import { playerLink, probClass, sparkCell, strokesCell, wireSparkTips } from "./cells.js";
import { DROP_CLASSES, dropIn, dropInPrep } from "./sim.js";
import { loadProj } from "./projection.js";

/* Simulated seasons per slider move. 8,000 costs ~50ms on a laptop and lands
   the run-to-run spread near half a point, which is the same noise floor the
   2026 replay carries and the reason both round to the whole percent above
   10%. Ten times as many buys about a third of a point and a visible stutter,
   which is the wrong trade for a control you drag. */
const DRAWS = 8000;

/* One slider per class a reader can choose to enter, in calendar order of
   how much a start there is worth. Playoffs are deliberately absent: they are
   won, not entered, and watching them appear as you drag the others is the
   most interesting thing on the tab. */
const SLIDERS = [
  ["elite", "DGPT", "Elite Series stops"],
  ["elite_plus", "DGPT+", "the bigger Elite Series stops, worth a third more"],
  ["major", "Majors", "worth twice an Elite Series stop; only your best two count"],
  ["jomez", "JomezPro", "bonus points only, and every one of them counts"],
];

const BANNER = `Nobody here is real. The 2027 calendar is announced, so the season
  you are entering is — but the player entering it is one you invented, and the field
  is <b>this season's players at this season's ratings</b>, carried forward unchanged,
  under the assumption that 2027's points structure and playoff ladder are 2026's.
  Then the hardest part: the field does not play you back. You are added to the
  projected standings without taking a single point off anyone in them, so at the very
  top you are racing a player the model still has finishing where they were.`;

/* ---------- the state a reader is editing ---------- */

/* Somebody on the bubble, which is the one place on this table where every
   column has something to say: a rating that finishes near the automatic-bid
   cut, and the schedule the model projects for them. Anywhere higher and the
   odds are all 100%; anywhere lower and they are all zero. */
function bubbleCfg(d) {
  const cut = d.meta.cut;
  let who = d.players[0];
  for (const p of d.players) {
    if (Math.abs(p.mean_rank - cut) < Math.abs(who.mean_rank - cut)) who = p;
  }
  const clsOf = new Map(d.schedule.filter((e) => e.ei != null).map((e) => [e.ei, e.cls]));
  const cfg = { rating: who.rating, who: who.name };
  for (const cls of DROP_CLASSES) cfg[cls] = 0;
  who.att.forEach((a, ei) => {
    const cls = clsOf.get(ei);
    if (cls in cfg) cfg[cls] += a;
  });
  for (const cls of DROP_CLASSES) cfg[cls] = Math.round(cfg[cls]);
  return cfg;
}

const cfgFor = (d) => (state.whatif[state.div] ||= bubbleCfg(d));

/* Projected points to a tenth. The bundle rounds every published mean_pts the
   same way, and 8,000 seasons cannot resolve a hundredth of a point anyway —
   left raw this reads "872.42" beside a column of "925.2"s. */
const pts1 = (x) => Math.round(x * 10) / 10;

/* ---------- the controls ---------- */

/* One slider. The explanation of what a class is worth rides on the "of N"
   mark rather than in a column of its own — it is a fact about the points
   table, not about the number you picked, and it buys back the row. */
function sliderHtml(key, label, note, value, max) {
  return `<div class="wi-row">
    <label class="wi-label" for="wi-${key}">${label}
      <span class="wi-of" ${tipAttrs(note)}>of ${max}</span></label>
    <input class="wi-range" type="range" id="wi-${key}" data-key="${key}"
      min="0" max="${max}" step="1" value="${value}"
      aria-describedby="wi-${key}-out">
    <output class="wi-out" id="wi-${key}-out">${value}</output>
  </div>`;
}

/* Where a rating would sit among the players the 2027 table lists. Rebuilt on
   every drag, so it is a function rather than a string. */
function seatText(d, rating) {
  const above = d.players.filter((p) => p.rating > rating).length;
  return above === 0
    ? "the best rating in the field"
    : `${ordinal(above + 1)}-best rating of the ${d.players.length} the 2027 table lists`;
}

function controlsHtml(d, prep, cfg) {
  const m = d.meta, [lo, hi] = prep.range;
  const rating = Math.min(hi, Math.max(lo, cfg.rating));
  const ratingTip = `PDGA player rating — the model's only input for how well somebody
    plays. About ${m.rating_pts_per_stroke} rating points are worth a stroke a round in
    ${m.division}, which is why this slider moves the numbers below so hard.`;
  const rows = SLIDERS.map(([key, label, note]) =>
    sliderHtml(key, label, `${label} — ${note}`, cfg[key],
      prep.plan.find((s) => s.cls === key).list.length)).join("");
  return `<div class="pv-panel">
    <div class="pv-head"><h2>Your 2027</h2>
      <span class="pv-form">${m.division} · ${m.roster_size} rivals</span></div>
    <div class="pv-body">
      <div class="wi-controls">
        <div class="wi-row wi-rating">
          <label class="wi-label" for="wi-rating">Rating
            <span class="wi-of" ${tipAttrs(ratingTip)}>PDGA</span></label>
          <input class="wi-range" type="range" id="wi-rating" data-key="rating"
            min="${lo}" max="${hi}" step="1" value="${rating}" aria-describedby="wi-rating-out">
          <output class="wi-out" id="wi-rating-out">${rating}</output>
          <span class="wi-hint" id="wi-seat">${seatText(d, rating)}</span>
        </div>
      </div>
      <div class="wi-controls wi-2up">${rows}</div>
      <p class="pv-note"><b>Which</b> events is the model's to pick — a fresh random draw
        of that many from each class every simulated season. The two playoff events are
        not on the list: you qualify for those on points or you do not, which is part of
        what is being computed below, from
        ${DRAWS.toLocaleString()} simulated seasons each time you move a slider.</p>
      <p class="pv-note"><button class="btn" id="wi-reset">reset to the bubble</button>
        <span class="wi-hint">the starting position is ${cfg.who}'s projected 2027 —
          the season the model has finishing nearest the automatic-bid cut, which is the
          one place every column below has something to say</span></p>
    </div>
  </div>`;
}

/* ---------- the readout ---------- */

/* Every column the 2027 table carries, with the copy it carries them under —
   a reader who has learned that table should not have to learn a second set
   of names here. Odds first, because they are the question. */
function tiles(m) {
  const pct = (v) => `<b class="${probClass(v)}">${fmtPct(v)}</b>`;
  return [
    ["p_champ", "Cup", `P(in the ${m.championship} field): an automatic bid, a ${m.playoff2} qualifier, or an event win`, pct, true],
    ["p_cup_win", "Win Cup", `P(wins the ${m.championship}): the ${m.cup_rounds}-round championship played out from your seed's starting strokes. A race — the whole field's odds add to 100%`, pct, true],
    ["p_cut", "Auto Bid", `P(finish top ${m.cut} in the 2027 standings — an automatic Cup berth)`, pct, true],
    ["p_perf", "KCWO Bid", `P(you take one of the ${m.perf_spots} Cup places won at the ${m.playoff2}, from outside the standings cut)`, pct, false],
    ["p_play1", m.playoff1, `P(you make the ${m.playoff1} field — top ${m.play1_cut} on points, expanding to ${m.play1_fill} if it doesn't fill)`, pct, false],
    ["p_play2", "KCWO", `P(you make the ${m.playoff2} field — top ${m.play2_cut} on points, plus the top ${m.play2_perf} ${m.playoff1} finishers from outside it)`, pct, false],
    ["p_first", "Wins it", "P(you finish first in the 2027 World Standings)", pct, false],
    ["mean_pts", "Proj. pts", "The points total you finish on, averaged over every simulated season", (v) => `<b>${fmtPts(pts1(v))}</b>`, false],
    ["mean_rank", "Proj. rank", "Your average finishing position in the projected standings", (v) => `<b>${v.toFixed(1)}</b>`, false],
    ["exp_starts", "Proj. starts", "Events you play, the two playoff events included when you earn them", (v) => `<b>${v.toFixed(1)}</b>`, false],
  ];
}

function tilesHtml(m, r) {
  return `<div class="wi-tiles">${tiles(m).map(([key, label, note, fmt, big]) =>
    `<div class="wi-tile${big ? " wi-big" : ""}" data-key="${key}">
      <span class="wi-tile-v">${fmt(r[key])}</span>
      <span class="wi-tile-k" ${tipAttrs(note)}>${label}</span>
    </div>`).join("")}</div>`;
}

/* The two distributions, drawn by the same code that draws them in every
   table row — a synthetic player with no PDGA number, which is the one thing
   about this reader that is definitely true. */
function distHtml(m, r) {
  const me = { pdga: 0, hist: r.hist, strokes: r.strokes };
  return `<div class="wi-dists">
    <div>
      <p class="band">Where you finish</p>
      ${sparkCell(me, m)}
      <p class="hint">One bar per position across all ${r.draws.toLocaleString()} seasons,
        green inside the top ${m.cut}, a last bar for ${m.max_hist_rank}th or worse.
        Tap or hover for the exact odds.</p>
    </div>
    <div>
      <p class="band">What you'd start the Cup on</p>
      ${strokesCell(me, m)}
      <p class="hint">Your finishing position run through the seed ladder
        (${fmtStroke(m.start_strokes.values[0])} for the No. 1 seed, E for the bottom
        seeds), with a final bar for the seasons you never reach the field.</p>
    </div>
  </div>`;
}

/* Who you would have landed between. The table above is abstract until it has
   a name in it, and this is the cheapest way to put one there. */
function neighboursHtml(d, r) {
  const near = [...d.players].sort((a, b) => b.mean_pts - a.mean_pts);
  let at = near.findIndex((p) => p.mean_pts < r.mean_pts);
  if (at < 0) at = near.length;
  const rows = [];
  // three above, you, three below: `at` is the first player you outscore, so
  // the same index runs both halves and only the marker moves
  for (let i = at - 3; i < at + 3; i++) {
    if (i === at) {
      rows.push(`<tr class="wi-you"><td>you</td><td class="num"><b>${fmtPts(pts1(r.mean_pts))}</b></td>
        <td class="num"><b class="${probClass(r.p_champ)}">${fmtPct(r.p_champ)}</b></td></tr>`);
    }
    const p = near[i];
    if (!p) continue;
    rows.push(`<tr><td>${playerLink(p)} <span class="dim">${p.rating}</span></td>
      <td class="num">${fmtPts(p.mean_pts)}</td>
      <td class="num"><span class="${probClass(p.p_champ)}">${fmtPct(p.p_champ)}</span></td></tr>`);
  }
  return `<div>
    <p class="band">Who you'd have landed between</p>
    <table class="table-ledger detail-tbl"><thead><tr>
      <th>Player</th><th class="num">Proj. pts</th><th class="num">Cup</th>
    </tr></thead><tbody>${rows.join("")}</tbody></table>
    <p class="hint">Ordered by projected points, so this is where the model would file
      you — remembering that it has not taken those points off anybody.</p>
  </div>`;
}

const NOTES = [
  `<b>The field is frozen.</b> You are added to the 2027 standings, not swapped into
   them: every point you win here is a point nobody else loses. That flatters you
   most where the race is tightest, and at the very top it is stranger than
   flattering — ask for the best rating in the field and you are racing somebody the
   model still has winning the season.`,
  `<b>Which events is a coin, not a plan.</b> Each simulated season draws a fresh
   random subset of the size you asked for. A player who only enters the events that
   suit them would do better than this; one who enters the four toughest would do
   worse.`,
  `<b>The doubles championship is not on the sliders.</b> It needs a partner, and 2027
   pairings cannot be known, so it is left out of your season entirely rather than
   guessed at.`,
  `<b>The Cup's opponents are averages.</b> Win Cup plays the championship out against
   the mean rating at each seed rather than against a distribution over who holds it,
   which narrows the spread of fields you could meet there.`,
  `<b>Everything the 2027 tab assumes, this tab assumes too</b> — the ratings, the
   attendance model behind the field sizes, the 2026 points curves and counting caps,
   and a roster with no 2027 rookie in it.`,
];

const notesHtml = () => `<details class="expl">
  <summary>What this what-if assumes</summary>
  <div class="expl-body">
    <p class="expl-lede">On top of everything the 2027 projection already assumes —
      that list is on the 2027 tab, and the last item below is a reminder that it still
      applies — dropping a player into that field costs the approximations here. None of
      them is hidden in the code.</p>
    <ol class="expl-list">${NOTES.map((n) => `<li>${n}</li>`).join("")}</ol>
  </div>
</details>`;

/* ---------- the view ---------- */

let prep = null;      // per-bundle, rebuilt when the division changes
let prepKey = null;
let pending = 0;      // rAF handle: a drag coalesces into one run per frame

function recompute(d, cfg) {
  const m = d.meta;
  const r = dropIn(prep, cfg, DRAWS);
  $("#wi-results").innerHTML =
    tilesHtml(m, r) +
    `<div class="wi-split">${distHtml(m, r)}${neighboursHtml(d, r)}</div>
     <p class="exact-odds">Unrounded: ${[["Cup", r.p_champ], ["Win Cup", r.p_cup_win],
       ["Auto Bid", r.p_cut], ["KCWO Bid", r.p_perf]]
       .map(([k, v]) => `${k} <b>${fmtPctExact(v)}</b>`).join(" · ")}
       <span class="dim" ${tipAttrs(`Rounded above 10% because the model cannot resolve a tenth of a percent here: this runs ${DRAWS.toLocaleString()} seasons per move and carries about half a point of noise between runs.`)}>why these differ from the tiles</span></p>`;
  wireSparkTips($("#wi-results"), m);
}

function wire(el, d, cfg) {
  const queue = () => {
    if (pending) return;
    pending = requestAnimationFrame(() => { pending = 0; recompute(d, cfg); });
  };
  el.querySelectorAll(".wi-range").forEach((input) =>
    input.addEventListener("input", () => {
      cfg[input.dataset.key] = +input.value;
      el.querySelector(`#wi-${input.dataset.key}-out`).textContent = input.value;
      if (input.dataset.key === "rating") $("#wi-seat").textContent = seatText(d, cfg.rating);
      queue();
    }));
  el.querySelector("#wi-reset").addEventListener("click", () => {
    state.whatif[state.div] = bubbleCfg(d);
    renderWhatIf();
  });
}

async function renderWhatIf() {
  // A queued frame closes over the bundle and config it was queued with, and
  // this call is about to replace the element it would write into — which is
  // how a fast division switch used to land MPO's odds in the FPO panel.
  if (pending) { cancelAnimationFrame(pending); pending = 0; }
  const el = $("#view-whatif");
  const d = await loadProj("2027", state.div);
  $("#kicker").textContent = "2027 DGPT Season Projection";
  $("#headline").textContent = "What if you were in the 2027 field?";
  $("#how-link").href = "how-it-works.html#experimental";
  $("#live-note").hidden = true;
  $("#sub-lede").textContent = "Your odds, if you joined the projected 2027 season";

  if (d && (prepKey !== state.div || !prep)) {
    prep = dropInPrep(d);
    prepKey = state.div;
  }
  if (!d || !prep) {
    $("#stakes-line").innerHTML = `<b class="al">Not published yet.</b>`;
    $("#meta-line").textContent = "no bundle available";
    el.innerHTML = `<p class="pv-intro">This what-if has not been generated yet. It rides
      along with the 2027 projection on the daily refresh, and a bundle published before
      it existed carries no field to drop into; check back after the next one.</p>`;
    return;
  }

  const m = d.meta;
  const cfg = cfgFor(d);
  {
    const f = updatedAt(m);
    const mins = Math.max(0, Math.round((Date.now() - f.getTime()) / 60000));
    $("#meta-line").innerHTML = `field projected ${agoText(mins)} · ` +
      `${DRAWS.toLocaleString()} seasons per move · top ${m.cut} qualify directly, ` +
      `field of ${m.field_size}`;
  }
  $("#stakes-line").innerHTML = `<span class="exp-chip">Highly experimental</span> ` +
    `one invented ${m.division} player against ${m.roster_size} real ones`;

  el.innerHTML = `
    <div class="exp-banner">
      <p class="exp-title">A player who doesn't exist · no results · no registrations · the field doesn't play you back</p>
      <p class="exp-text">${BANNER}</p>
    </div>
    ${notesHtml()}
    ${controlsHtml(d, prep, cfg)}
    <div id="wi-results"></div>`;

  wire(el, d, cfg);
  recompute(d, cfg);
}

export { renderWhatIf };
